src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -243,39 +334,13 @@ private ValueTask ShutdownRequestExecutionQueueAsync()
     /// </summary>
     private async void JsonRpc_Disconnected(object? sender, JsonRpcDisconnectedEventArgs e)
     {
-        if (_shuttingDown)
-        {
-            // We're already in the normal shutdown -> exit path, no need to do anything.
-            return;
-        }
-
-        var message = $"Encountered unexpected jsonrpc disconnect, Reason={e.Reason}, Description={e.Description}, Exception={e.Exception}";
-        _logger.LogWarning(message);
-
-        var lspServices = GetLspServices();
-        var lifeCycleManager = lspServices.GetRequiredService<ILifeCycleManager>();
-
-        await lifeCycleManager.ShutdownAsync(message).ConfigureAwait(false);
-        await lifeCycleManager.ExitAsync().ConfigureAwait(false);
+        // It is possible this gets called during normal shutdown and exit.
+        // ShutdownAsync and ExitAsync will no-op if shutdown was already triggered by something else.
+        await ShutdownAsync(message: "Shutdown triggered by JsonRpc disconnect").ConfigureAwait(false);
+        await ExitAsync().ConfigureAwait(false);
     }
 #pragma warning disable VSTHRD100
 
-    /// <summary>
-    /// Disposes the LanguageServer, clearing and shutting down the queue and exiting.
-    /// Can be called if the Server needs to be shut down outside of 'shutdown' and 'exit' requests.
-    /// </summary>
-    public async ValueTask DisposeAsync()
```

Comment:
having dispose along with all the rest just made our brains hurt.  So deleted.

Reply:
nice!